plugins/meta/vrf/main.go

```diff
@@ -0,0 +1,183 @@
+	"github.com/containernetworking/plugins/pkg/ns"
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+)
+
+// VRFNetConf represents the vrf configuration.
+type VRFNetConf struct {
+	types.NetConf
+
+	// VRFName is the name of the vrf to add the interface to.
+	VRFName string `json:"vrfname"`
+}
+
+func main() {
+	skel.PluginMain(cmdAdd, cmdCheck, cmdDel, version.PluginSupports("0.4.0"), bv.BuildString("vrf"))
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	conf, result, err := parseConf(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	if conf.PrevResult == nil {
+		return fmt.Errorf("missing prevResult from earlier plugin")
+	}
+
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		vrf, err := findVRF(conf.VRFName)
+
+		if _, ok := err.(netlink.LinkNotFoundError); ok {
+			vrf, err = createVRF(conf.VRFName)
+		}
+
+		if err != nil {
+			return err
+		}
+
+		err = addInterface(vrf, args.IfName)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("cmdAdd failed: %v", err)
+	}
+
+	if result == nil {
+		result = &current.Result{}
+	}
+
+	return types.PrintResult(result, conf.CNIVersion)
+}
+
+func cmdDel(args *skel.CmdArgs) error {
+	conf, _, err := parseConf(args.StdinData)
+	if err != nil {
+		return err
+	}
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		vrf, err := findVRF(conf.VRFName)
+		if err != nil {
+			return err
+		}
+
+		err = resetMaster(args.IfName)
+		if err != nil {
+			return err
+		}
+
+		interfaces, err := assignedInterfaces(vrf)
+		if err != nil {
+			return err
+		}
+
+		// Meaning, we are deleting the last interface assigned to the VRF
+		if len(interfaces) == 0 {
+			err = netlink.LinkDel(vrf)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("cmdDel failed: %v", err)
+	}
+	return nil
+}
+
+func cmdCheck(args *skel.CmdArgs) error {
+	conf, _, err := parseConf(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	// Ensure we have previous result.
+	if conf.PrevResult == nil {
+		return fmt.Errorf("missing prevResult from earlier plugin")
+	}
+
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		vrf, err := findVRF(conf.VRFName)
+		if err != nil {
+			return err
+		}
+		vrfInterfaces, err := assignedInterfaces(vrf)
+
+		found := false
+		for _, intf := range vrfInterfaces {
+			if intf.Attrs().Name == args.IfName {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return fmt.Errorf("Failed to find %s associated to vrf %s", args.IfName, conf.VRFName)
+		}
+		return nil
+	})
+
+	return nil
+}
+
+func parseConf(data []byte) (*VRFNetConf, *current.Result, error) {
+	conf := VRFNetConf{}
+	if err := json.Unmarshal(data, &conf); err != nil {
+		return nil, nil, fmt.Errorf("failed to load netconf: %v", err)
+	}
+
+	if conf.VRFName == "" {
+		return nil, nil, fmt.Errorf("configuration is expected to have a valid vrf name")
+	}
+
+	// Parse previous result.
```

Comment:
Hmmm... Same comment is below, so this looks a bit suspicious as real parsing is done later, while there is only checking if it was provided in raw form.

Reply:
Ouch

Reply:
At least I was consistent in leaving the comments from where I copied the structure :laughing:  https://github.com/containernetworking/plugins/blob/0a1421a08cc909aa67360f749e90bab60206a01f/plugins/meta/firewall/firewall.go#L75

Reply:
Done